docs/virtual_machines/accessing_virtual_machines.md

```diff
@@ -2,184 +2,133 @@
-some other options that grant more flexibility.
+SSH public keys into the virtual machines at boot. This allows automation tools
+(like Ansible) to provision the virtual machine. It also gives operators a way
+of gaining secure and passwordless access to a virtual machine.
 
-The VM's access credential api allows statically injecting ssh public keys at creation
-time independently of the cloud-init user data by placing the ssh public key
-in a Kubernetes secret. This is useful because it allows people creating virtual
-machines to separate the application data in their cloud-init user data from the
-credentials used to access the virtual machine.
+KubeVirt provides multiple ways to inject SSH public keys into a virtual
+machine. In general, these methods fall into two categories.
+[Static key injection](./accessing_virtual_machines.md#static-ssh-key-injection-via-cloud-init),
+which places keys on the virtual machine the first time it is booted, and
+[dynamic injection](./accessing_virtual_machines.md#dynamic-ssh-key-injection-via-qemu-user-agent),
+which allows keys to be dynamically updated both at boot and during runtime.
 
-For example, someone can put their ssh key into a Kubernetes secret like this.
+Once a SSH public key is injected into the virtual machine, access to it can be
+gained with the help of `virtctl`.
 
-```
-# Place ssh key into a secret
+### Static SSH public key injection via cloud-init
 
-kubectl create secret generic my-pub-key --from-file=key1=/id_rsa.pub
+Users creating virtual machines have the ability to provide startup scripts to
+their virtual machines, which allow any number of customization operations to
+take place. Placing SSH public keys into a
+[cloud-init startup script](./startup_scripts.md#cloud-init)
+is one option for injecting them into a virtual machine. However, there are
+other, more flexible options available.
+
+The virtual machine's access credential API allows statically injecting SSH
+public keys at startup time independently of the cloud-init user data by
+placing the SSH public key into a Kubernetes `Secret`. This allows to keep the
+application data in the cloud-init user data separated from the credentials
+used to access the virtual machine.
+
+A Kubernetes `Secret` can be created from an SSH public key like this:
+
+```shell
+# Place SSH public key into a Secret
+kubectl create secret generic my-pub-key --from-file=key1=id_rsa.pub
 ```
 
-Then assign that key to the virtual machine with the access credentials api
-using the configDrive propagation method. Note here how the cloud-init user
-data is not touched. KubeVirt is injecting the ssh key into the virtual machine
-using the machine generated cloud-init metadata, and not the user data. This
-keeps the application user date separate from credentials.
+The `Secret` containing the public key is then assigned to a virtual machine
+using the access credentials API with the `configDrive` propagation method.
```

Comment:
@qinqon I remember discussing w/ you some time ago; does it really need to be `configDrive` ? I have the slight recollection you told me this would also work w/ `NoCloud` . 

Did I misunderstand you / am I wrong ?

Reply:
It's only implemented for `configDrive` at the moment.

See variable `keys` that is injected into the config drive metadata:
https://github.com/kubevirt/kubevirt/blob/4de7e9773109226c1ecc9e4a0df9ed2f1f6b0cd5/pkg/cloud-init/cloud-init.go#L137

Reply:
Thanks. Looking at the code and at the [current README](https://kubevirt.io/user-guide/virtual_machines/startup_scripts/#injecting-ssh-keys-with-cloud-inits-cloud-config) I think I misunderstood: we can sneak the public keys via secret via the `userData` which is defined in a k8s secret. 

Thanks for the code link - the truth (as always ...) is in the code.

Reply:
That's weird this is working for us on capk, https://github.com/kubernetes-sigs/cluster-api-provider-kubevirt/pull/185/files#diff-bab123ad42a3d34ff93b21e3c9695fb2d81b2ea4bcd87eaea42b987565153bdaR152